chore(refactor): Operator-sdk upgrade

RealHarshThakur opened the pull request:
## Pull Request template

**Why is this PR required? What issue does it fix?**:
Operator-sdk is used to scaffold the project and beyond that, it will help when creating webhooks, etc

**What this PR does?**:
Upgrades operator-sdk to latest version

**Does this PR require any upgrade changes?**:

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** : 
_Mention if this PR is part of any design or a continuation of previous PRs_


Pending tasks:
- [x] Migrate project structure
- [x] Fix import types
- [x] Upgrade client-go
- [x] Fix ndm operator's main.go (it has outdated operator-sdk dependencies)
- [x] Fix unit tests 
- [x] Fix integration tests which are failing due to lack of preupgrade checks 



References:
* https://github.com/kubernetes-sigs/controller-runtime/releases (for breaking changes in controller runtime)
* https://sdk.operatorframework.io/docs/building-operators/golang/migration/ 


**Checklist:**
- [ ] Fixes #<issue number>
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

akhilerm commented on Dockerfile:
```diff
@@ -0,0 +1,44 @@
+# */
+
+
+# Build the manager binary
```
What  is this `Dockerfile` used for?

akhilerm commented on PROJECT:
```diff
@@ -0,0 +1,28 @@
+    namespaced: true
+  controller: true
+  domain: openebs.io
+  group: blockdevice
```
The group is `openebs.io` right? This will resolve to `blockdevice.blockdevice.openebs.io`

akhilerm commented on apis/blockdevice/v1alpha1/groupversion_info.go:
```diff
@@ -14,29 +14,23 @@ See the License for the specific language governing permissions and
-	// SchemeGroupVersion is group version used to register these objects
-	SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: APIVersion}
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "blockdevice.openebs.io", Version: "v1alpha1"}
```
Shouldn't this be just `openebs.io`?

akhilerm commented on .dockerignore:
```diff
@@ -0,0 +1,5 @@
+# More info: https://docs.docker.com/engine/reference/builder/#dockerignore-file
```
Why was this file added?

akhilerm commented on apis/blockdevice/v1alpha1/blockdevice_types.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 // Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
```
Is this required now?

akhilerm commented on apis/blockdeviceclaim/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "blockdeviceclaim.openebs.io", Version: "v1alpha1"}
```
Same comment as above on the group name.

akhilerm commented on build/boilerplate.go.txt:
```diff
@@ -0,0 +1,15 @@
+/*
```
This file is already present at https://github.com/openebs/node-disk-manager/blob/master/build/custom-boilerplate.go.txt

akhilerm commented on cmd/manager/main.go:
```diff
@@ -14,156 +14,156 @@ See the License for the specific language governing permissions and
-	v041_v042UpgradeTask := v041_042.NewUpgradeTask("0.4.1", "0.4.2", client)
-	return upgrade.RunUpgrade(v040_v041UpgradeTask, v041_v042UpgradeTask)
-}
+// package main
```
Why is the complete file commented out?

RealHarshThakur commented:
@akhilerm  Mentioned some of these concerns in Slack too. 
I am not sure about some of the changes:
 * Changed GVK (`blockdevice.openebs.io` to `blockdevices.blockdevice.openebs.io`) . Should probably revert back to original API group . 
 *   I don't know if this is the right way to do this: https://github.com/openebs/node-disk-manager/pull/561/files#diff-a20906eb0434b5deaf080095f98faa6ab0cd1420cece8129fc0ec7102c14d9f6R278
 *   Currently ndm operator main.go is commented because it has outdated dependencies of operator-sdk , I am wondering if we still need ndm-operator at all. Anyway, updated the rest of the code base ndm-operator relies on 
 
 Boilerplate, dockerfiles, etc were generated by operator-sdk which I'll clean up and structure in Makefile when operator-sdk is to be used.

akhilerm commented on apis/blockdevice/v1alpha1/crds.go:
```diff
@@ -26,7 +26,7 @@ const (
+	BlockDeviceResourceName = BlockDeviceResourcePlural + "." + GroupVersion.Group
 
 	// BlockDeviceClaimResourceKind is the kind of block device claim CRD
 	BlockDeviceClaimResourceKind = "BlockDeviceClaim"
```
Why do we need this variables in `blockdevice/crds`. It should be in `blockdeviceclaims/crds` right?

akhilerm commented on config/certmanager/certificate.yaml:
```diff
@@ -0,0 +1,25 @@
+# The following manifests contain a self-signed issuer CR and a certificate CR.
```
Why are the files in `config/` directory required?

akhilerm commented on cmd/ndm_daemonset/controller/controller.go:
```diff
@@ -275,10 +275,13 @@ func (c *Controller) Broadcast() {
 
 // run waits until it gets any interrupt signals
-func (c *Controller) run(threadiness int, stopCh <-chan struct{}) error {
+func (c *Controller) run(threadiness int, ctx context.Context) error {
```
We are passing channel from [here](https://github.com/openebs/node-disk-manager/pull/561/files#diff-a20906eb0434b5deaf080095f98faa6ab0cd1420cece8129fc0ec7102c14d9f6R259), how will the context work in this case?

akhilerm commented on controllers/blockdevice/blockdevice_controller.go:
```diff
@@ -145,9 +119,16 @@ func (r *ReconcileBlockDevice) Reconcile(request reconcile.Request) (reconcile.R
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apis.BlockDevice{}).
+		Owns(&apis.BlockDevice{}).
+		Complete(r)
```
What is the use of this function?

shubham14bajpai replied:
It is equivalent to Build() functions we have in builder pattern

akhilerm commented on deploy/crds/blockdevice.openebs.io_blockdevices.yaml:
```diff
@@ -6,9 +6,9 @@ metadata:
     controller-gen.kubebuilder.io/version: v0.4.0
   creationTimestamp: null
-  name: blockdevices.openebs.io
+  name: blockdevices.blockdevice.openebs.io
```
Is there any way to change it and use it like how we previously had?

akhilerm commented on go.mod:
```diff
@@ -1,29 +1,27 @@
+	golang.org/x/sys v0.0.0-20200622214017-ed371f2e16b4
+	google.golang.org/grpc v1.27.0
+	google.golang.org/protobuf v1.24.0
+	k8s.io/api v0.19.2
```
Can the k8s version be at 1.20?

akhilerm commented on main.go:
```diff
@@ -0,0 +1,116 @@
+/*
```
Why do we have the `main.go` in the root of the project?

akhilerm commented on pkg/crds/build.go:
```diff
@@ -20,7 +20,7 @@ import (
-	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 )
 
```
This file can be deleted as we are no longer installing from code.

akhilerm reviewed (commented):
Some comments:
1. What is the use of `config/` directory
2. Are the controllers meant to be at the root of the project?

shubham14bajpai commented on config/crd/patches/webhook_in_blockdeviceclaims.yaml:
```diff
@@ -0,0 +1,14 @@
+# The following patch enables a conversion webhook for the CRD
```
All the auto-generated configs are not needed here, we can remove things like webhooks

shubham14bajpai commented on config/default/kustomization.yaml:
```diff
@@ -0,0 +1,74 @@
+# Adds namespace to all resources.
```
I think we can remove this as we don't use kustomize for deployment

shubham14bajpai reviewed (commented):
For the packaging of the latest operator-sdk the main.go, apis/ and controller/ are at the root which is ideal for a single controller. But in this repo we have multiple operators so can we move them back into cmd/ and pkg/ structure instead of strictly following the operator-sdk pattern.

RealHarshThakur commented:
@shubham14bajpai I think it's possible to have multiple controllers with same layout because main.go creates a `manager` to run all of them.  We can remove the kustomize and others thing we don't use but I am not sure if we should move away from project structure as it will make upgrading the operator-sdk version difficult as mentioned in their docs.  https://v1-5-x.sdk.operatorframework.io/docs/faqs/#can-i-customize-the-projects-generated-with-sdk-tool

akhilerm commented:
> @shubham14bajpai I think it's possible to have multiple controllers with same layout because main.go creates a `manager` to run all of them. We can remove the kustomize and others thing we don't use but I am not sure if we should move away from project structure as it will make upgrading the operator-sdk version difficult as mentioned in their docs. It will also hinder using some of the features https://v1-5-x.sdk.operatorframework.io/docs/faqs/#can-i-customize-the-projects-generated-with-sdk-tool

How is it done in jiva operator @shubham14bajpai ?

shubham14bajpai commented:
> > @shubham14bajpai I think it's possible to have multiple controllers with same layout because main.go creates a `manager` to run all of them. We can remove the kustomize and others thing we don't use but I am not sure if we should move away from project structure as it will make upgrading the operator-sdk version difficult as mentioned in their docs. It will also hinder using some of the features https://v1-5-x.sdk.operatorframework.io/docs/faqs/#can-i-customize-the-projects-generated-with-sdk-tool
> 
> How is it done in jiva operator @shubham14bajpai ?

In case of jiva-operator we wanted to be less dependent on the operator-sdk binary as that itself has breaking changes if we upgrade after a long time. As the operator-sdk generates the code using the standard k8s kubebuilder etc. I restructured the packages in line with other repos and added targets to the makefile using standard k8s patterns as I did for the crd generation PR. This helps us have more control over smaller parts of the code.

RealHarshThakur commented:
Currently, kubebuilder and operator-sdk are following the same package structure. Also, I think there are some good features we have to start utilizing from Operator-sdk like OLM, score card testing and small features like [this](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) which might be valuable at some point.  After their 1.0 release, there are no [breaking changes](https://github.com/operator-framework/operator-sdk/releases).  Having said that, if we're sure we don't need those features , then I think we can stop relying on operator-sdk altogether but  IMHO, I think we should stick to Operator-SDK and if we ever face any issue, we can easily fall back on controller-gen, etc .

shubham14bajpai commented:
Yes it makes sense if these features are used. In case of jiva-operator it was a simple reconciler and these features are not required.

akhilerm commented:
@RealHarshThakur Are there any blockers for this PR ?

RealHarshThakur commented:
@akhilerm Yes
* Integration tests are failing :  We need to add preupgrade checks in operator but since we're not using some of the packages we used to previously, have to find a new way. 
* Unit test is failing on Block device for some reason. 

Otherwise, the PR is ready

akhilerm commented:
@RealHarshThakur  Unit tests are failing because of [this](https://github.com/openebs/node-disk-manager/pull/561/checks?check_run_id=2375181157#step:4:35). Can you check if the operator SDK upgrade had brought in changes?

RealHarshThakur commented:
@akhilerm  I tried to debug but couldn't figure out immediately why it's happening. I need to try again. I am not sure if it's because of operator-sdk or that change in `SetupHandler` which returned context instead of channel . Would really appreciate if you took a look at it once

akhilerm commented:
@RealHarshThakur can you update on this PR?

TomMD commented on vendor/github.com/Azure/go-autorest/autorest/date/timerfc1123.go:
```diff
@@ -0,0 +1,100 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
It appears you'll want to add to `.muse/config.toml` a line such as:

```
ignoreFiles = """
vendor/**
"""
```

RealHarshThakur replied:
ignore

Z0Marlin commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
+	patch := client.MergeFromWithOptions(blockDevice.DeepCopy())
 
-	err = c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	err = c.Clientset.Patch(context.TODO(), &blockDevice, patch)
```
If I'm not wrong `UpdateBlockDevice` is used to update bd with the new values provided in the `blockDevice` parameter right? But since we are using a copy of the `blockDevice` as the base in `MergeFromWithOptions` and then sending the same `blockDevice` in `ClientSet.Patch` won't the resultant patch have no changes? I could be wrong about the flow here. Let me know if I'm missing something.

RealHarshThakur replied:
I'm not sure about the best way to pick up changes here. If we use `Update` call then that results in a conflict. Even after I fetch it right before making the `Update` call . All the test cases are failing because of this, the only way I found which solves this is in the last commit (Fix ZFS test case) .

z0marlin replied:
Interesting. I tried debugging the `TestUpdateDevice`. I found that the final patch data being generated is `[]byte{123,125}`, which corresponds to `{}`. So looks like the device won't be updated since the patch is empty. The test also doesn't seem to be checking for this. I think the tests need to be updated too. Here's a screenshot of the debugging session (focus on the value of `data`):
![Screenshot from 2021-05-20 15-02-40](https://user-images.githubusercontent.com/31250892/118955788-a6c5db80-b97c-11eb-89da-06e0a795adac.png)

akhilerm replied:
@RealHarshThakur I think the concern raised by @z0marlin will happen. the patch is not having any changes. That is the reason the call is not erroring out.

akhilerm replied:
@RealHarshThakur Can you revisit the logic for the Update method, we need to find a suitable method to do it. I will also test it out locally.

akhilerm replied:
>  Also @z0marlin there will be changes in the patch in some cases, because the base object and etcd can differ. (though not always)

I think the above statement from me is wrong. If using the same object in the patch call and patch, there will be no changes. We have to see how a 3 way merge patch works.

z0marlin replied:
True @akhilrm. Going through the merge patch code, even I suspect the patch data will always be empty since the patch is by comparing the base object set in the `MergeFromWithOptions` call, and the object sent in the `Patch` (the second parameter). These two are same since the former is a deepcopy of the latter in this code.

RealHarshThakur replied:
@akhilerm Worst case scenario: I think we should `Update` or `Patch` the resource in the same function and remove/minimize the use of this function itself.  I did that in last commit (ZFS test case). WDYT?

akhilerm commented on Makefile:
```diff
@@ -71,7 +71,7 @@ export BASEIMAGE
 
 ifeq (${IMAGE_ORG}, )
-  IMAGE_ORG="openebs"
+  IMAGE_ORG="harshthakur"
```
instead of changing here, when running the command you can use like this
```console
make docker.ndm IMAGE_ORG="harshthakur"
```

akhilerm commented on PROJECT:
```diff
@@ -0,0 +1,23 @@
+domain: openebs.io
```
What is this file used for? Is it mandatory to include in the project?

RealHarshThakur replied:
This is used by Kubebuilder and Operator-SDK. They expect this to be there

akhilerm commented on api/v1alpha1/blockdevice_types.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The OpenEBS Authors
+Copyright 2021.
```
[This](https://github.com/openebs/node-disk-manager/blob/master/build/custom-boilerplate.go.txt) is the license header that should be used. You can edit the file to 2021.

akhilerm commented on api/v1alpha1/blockdevice_types.go:
```diff
@@ -254,7 +228,27 @@ const (
+// BlockDevice is the Schema for the blockdevices API
+// +kubebuilder:printcolumn:name="NodeName",type="string",JSONPath=`.spec.nodeAttributes.nodeName`
+// +kubebuilder:printcolumn:name="Path",type="string",JSONPath=`.spec.path`,priority=1
+// +kubebuilder:printcolumn:name="FSType",type="string",JSONPath=`.spec.filesystem.fsType`,priority=1
```
Shall we change this to deviceType as FSType will not always be present.

akhilerm commented on api/v1alpha1/blockdevice_types.go:
```diff
@@ -254,7 +228,27 @@ const (
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
```
Is a subresource required?

RealHarshThakur replied:
Yes. Basically, the desired state is expressed in `spec` and observed state is present in `status` sub resource . Client-go libraries can make use of this for updates, etc.

akhilerm commented on cmd/manager/main.go:
```diff
@@ -17,157 +17,98 @@ limitations under the License.
-	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
-	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	// to ensure that exec-entrypoint and run can make use of them.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Why is this import required?

RealHarshThakur replied:
I think this can be used for out-of-cluster authentication . So we can develop the package locally, just need to point it to a K8s cluster. Although, the disks detected, etc would be on local machine.

akhilerm replied:
But as far as I see the code, its not used. So it can be removed.

akhilerm replied:
Keep only the most basic things so that it works as previously and further changes can be added later.

akhilerm commented on api/v1alpha1/crds.go:
```diff
@@ -16,7 +16,7 @@ limitations under the License.
-const (
+var (
 	// BlockDeviceResourceKind is the kind of block device CRD
 	BlockDeviceResourceKind = "BlockDevice"
```
Except this and `BlockDeviceClaimResourceKind`, all others can be removed as they were used only during CRD installation from the ndm operator which is now removed.

akhilerm commented on api/v1alpha1/groupversion_info.go:
```diff
@@ -14,29 +14,23 @@ See the License for the specific language governing permissions and
-	// SchemeGroupVersion is group version used to register these objects
-	SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: APIVersion}
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "openebs.io", Version: "v1alpha1"}
```
Was the change from constants to using string literals directly done by the auto generated code?

RealHarshThakur replied:
It was boilter plate code IIRC

akhilerm commented on cmd/manager/main.go:
```diff
@@ -17,157 +17,98 @@ limitations under the License.
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
 
-func printVersion() {
-	klog.Infof("Go Version: %s", runtime.Version())
-	klog.Infof("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH)
-	klog.Infof("operator-sdk Version: %v", sdkVersion.Version)
-	klog.Infof("Version Tag: %s", version.GetVersion())
-	klog.Infof("Git Commit: %s", version.GetGitCommit())
```
Need to include this information while starting the operator so that we can know which version is being run.

akhilerm commented on cmd/manager/main.go:
```diff
@@ -17,157 +17,98 @@ limitations under the License.
-	defer r.Unset()
-
-	reconInterval := ReconciliationInterval
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
```
We had migrated from zap to klog in NDM. Therefore use only klog for logging.

RealHarshThakur replied:
This is a new library , not the orignal zap library:  	"sigs.k8s.io/controller-runtime/pkg/log/zap"

akhilerm replied:
okay. But the logger is different. We need to resort to using only one type of logger. Else the log formatting will be different from different code bases.

akhilerm commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -38,7 +39,9 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
 	blockDeviceCopy := blockDevice.DeepCopy()
-	err := c.Clientset.Create(context.TODO(), blockDeviceCopy)
+	err := c.Clientset.Create(context.TODO(), blockDeviceCopy, &client.CreateOptions{
+		FieldManager: "NDM",
```
Is this required on every create call? Also lower case is preferred.
```suggestion
		FieldManager: "ndm",
```

akhilerm commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
 	return nil
 }
 
+func boolPtr(b bool) *bool {
+	return &b
+}
+
```
This is not used

akhilerm commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
-		}
-	}
-
-	blockDeviceCopy = mergeBlockDeviceData(*blockDeviceCopy, *oldBlockDevice)
```
This function needs to be used for merging as we dont need to merge all the fields. Some fields are merged on condition depending on the status of the blockdevice.

akhilerm commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
-	err := c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	patch := client.MergeFrom(blockDevice.DeepCopy())
+	blockDevice.Status.State = NDMInactive
+	err := c.Clientset.Status().Patch(context.TODO(), &blockDevice, patch)
```
Is this `client.Status().` required because we have status as a subresource?

RealHarshThakur replied:
Yes

akhilerm replied:
Do we really need status as a subresource. Whats the best practice around that?

RealHarshThakur replied:
These are the conventions: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status

akhilerm commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
-	blockDeviceCopy := blockDevice.DeepCopy()
-	blockDeviceCopy.Status.State = NDMInactive
-	err := c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	patch := client.MergeFrom(blockDevice.DeepCopy())
```
Is this not giving out any errors?

RealHarshThakur replied:
Nope, why?

akhilerm commented on cmd/ndm_daemonset/controller/controller.go:
```diff
@@ -137,6 +137,8 @@ func NewController() (*Controller, error) {
 	}
 
+	s := mgr.GetScheme()
+	fmt.Println(s)
```
This needs to be removed before the final merge.

akhilerm commented on cmd/ndm_daemonset/probe/addhandler.go:
```diff
@@ -336,49 +338,68 @@ func (pe *ProbeEvent) deviceInUseByMayastor(bd blockdevice.BlockDevice, bdAPILis
+	bdAPI.Namespace = pe.Controller.Namespace
+
+	err := pe.Controller.Clientset.Create(context.TODO(), bdAPI.DeepCopy(), &client.CreateOptions{
+		FieldManager: "NDM",
```
```suggestion
		FieldManager: "ndm",
```

akhilerm commented on cmd/ndm_daemonset/probe/addhandler.go:
```diff
@@ -336,49 +338,68 @@ func (pe *ProbeEvent) deviceInUseByMayastor(bd blockdevice.BlockDevice, bdAPILis
+		return &bdAPI, false, err
+	}
 
-	err := pe.Controller.CreateBlockDevice(bdAPI)
```
Why is this change required? The blockdevices as part of ZFS localpv will only be updated by NDM. So there is no chance of conflicts there.

akhilerm commented on controllers/blockdevice/blockdevice_controller.go:
```diff
@@ -115,19 +89,19 @@ func (r *ReconcileBlockDevice) Reconcile(request reconcile.Request) (reconcile.R
 			// finalizer is not present, may be a BlockDevice claimed from previous release
 			instance.Finalizers = append(instance.Finalizers, controllerutil.BlockDeviceFinalizer)
-			err := r.client.Update(context.TODO(), instance)
+			err := r.Client.Update(context.TODO(), instance)
```
Arent these update calls causing an issue? like conflicts when updating in the NDM daemon.

akhilerm commented on cmd/manager/main.go:
```diff
@@ -17,157 +17,98 @@ limitations under the License.
+	if err = (&blockdeviceclaim.BlockDeviceClaimReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("BlockDeviceClaim"),
+		Scheme: mgr.GetScheme(),
```
The recorder is not initialized . It is needed for adding the events.

akhilerm commented on cmd/manager/main.go:
```diff
@@ -17,157 +17,98 @@ limitations under the License.
+	if err = (&blockdevice.BlockDeviceReconciler{
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("controllers").WithName("BlockDevice"),
+		Scheme: mgr.GetScheme(),
```
Here also recorder needs to be initialized.

akhilerm commented on controllers/blockdevice/blockdevice_controller.go:
```diff
@@ -140,13 +114,25 @@ func (r *ReconcileBlockDevice) Reconcile(request reconcile.Request) (reconcile.R
+func (r *BlockDeviceReconciler) updateBDStatus(state apis.DeviceClaimState, instance *apis.BlockDevice) error {
 	instance.Status.ClaimState = state
-	return r.client.Update(context.TODO(), instance)
+	err := r.Client.Update(context.TODO(), instance)
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return r.client.Update(context.TODO(), instance)
```

akhilerm commented on controllers/blockdeviceclaim/blockdeviceclaim_controller.go:
```diff
@@ -20,73 +20,52 @@ import (
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.2/pkg/reconcile
+func (r *BlockDeviceClaimReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
```
```suggestion
```

akhilerm commented on controllers/blockdeviceclaim/blockdeviceclaim_controller.go:
```diff
@@ -20,73 +20,52 @@ import (
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.2/pkg/reconcile
+func (r *BlockDeviceClaimReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	// _ = r.Log.WithValues("blockdeviceclaim", request.NamespacedName)
```
```suggestion
```

akhilerm commented on controllers/blockdeviceclaim/blockdeviceclaim_controller.go:
```diff
@@ -20,73 +20,52 @@ import (
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileBlockDeviceClaim) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// your logic here
```
```suggestion
```

akhilerm commented on hack/boilerplate.go.txt:
```diff
@@ -0,0 +1,15 @@
+/*
```
Instead of this file , use the `custom-boilerplate` in build directory.

akhilerm commented on go.mod:
```diff
@@ -1,29 +1,27 @@
 module github.com/openebs/node-disk-manager
 
-go 1.14
+go 1.15
```
When updating go version here, also update the go version used for building the binary in the correspnding `build/<component>/Dockerfile`

akhilerm commented on integration_tests/k8s/k8sapi.go:
```diff
@@ -118,7 +118,7 @@ func (c K8sClient) RestartPod(name string) error {
 	for pod := range pods {
 		if strings.Contains(pod, name) {
-			return c.ClientSet.CoreV1().Pods(Namespace).Delete(pod, &metav1.DeleteOptions{})
+			return c.ClientSet.CoreV1().Pods(Namespace).Delete(context.Background(), pod, metav1.DeleteOptions{})
```
```suggestion
			return c.ClientSet.CoreV1().Pods(Namespace).Delete(context.TODO(), pod, metav1.DeleteOptions{})
```

z0marlin replied:
I see `TODO` context being used at multiple places. Is there any root context we can replace these with?

akhilerm commented on integration_tests/k8s/k8sapi.go:
```diff
@@ -209,13 +209,13 @@ func (c K8sClient) DeleteClusterRoleBinding(clusterrolebinding rbacv1beta1.Clust
 // CreateCustomResourceDefinition creates a CRD
 func (c K8sClient) CreateCustomResourceDefinition(customResourceDefinition apiextensionsV1.CustomResourceDefinition) error {
-	_, err := c.APIextClient.ApiextensionsV1().CustomResourceDefinitions().Create(&customResourceDefinition)
+	_, err := c.APIextClient.ApiextensionsV1().CustomResourceDefinitions().Create(context.Background(), &customResourceDefinition, metav1.CreateOptions{})
```
`context.TODO()` needs to be used.

z0marlin replied:
I see `TODO` context being used at multiple places. Is there any root context we can replace these with?

akhilerm commented on pkg/setup/build_crd.go:
```diff
@@ -16,46 +16,46 @@ limitations under the License.
 
```
This file can be deleted.

akhilerm commented on pkg/setup/create_crd.go:
```diff
@@ -16,72 +16,72 @@ limitations under the License.
 
 package setup
 
```
Can be deleted.

akhilerm commented on pkg/setup/delete_crd.go:
```diff
@@ -16,36 +16,36 @@ limitations under the License.
 
 package setup
 
```
Can be deleted.

akhilerm commented on pkg/setup/install.go:
```diff
@@ -16,23 +16,23 @@ limitations under the License.
 
 package setup
 
```
Can be deleted.

akhilerm commented on pkg/upgrade/v040_041/preupgrade.go:
```diff
@@ -19,7 +19,7 @@ package v040_041
 import (
 	"context"
```
This package can be deleted as we are no longer using it.

akhilerm reviewed (changes requested):
@RealHarshThakur completed the review the requested some changes/ gave comments.

z0marlin commented on cmd/ndm_daemonset/controller/blockdevicestore.go:
```diff
@@ -83,52 +86,63 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
 // UpdateBlockDevice update the BlockDevice resource in etcd
 func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice, oldBlockDevice *apis.BlockDevice) error {
```
@akhilerm do we really need the second parameter in this function. From what I can understand, the second parameter `oldBlockDevice` is useful only when we want to change the name or namespace of an existing BD. Is this why we have kept it? I feel changing the name shouldn't be allowed by the update operation.

akhilerm replied:
The second parameter was provided because we were fetching the latest blockdevice and passing it along with the new resource that we wanted to update 

Name and namespace change is something that we don't want to do.

z0marlin replied:
I see. Since now we are anyways using another variable `tempBD` to store the fetched bd from etcd, I think we can do away with the second parameter?

z0marlin commented on pkg/select/blockdevice/filters_test.go:
```diff
@@ -18,8 +18,10 @@ package blockdevice
+
 	"github.com/stretchr/testify/assert"
 	"k8s.io/apimachinery/pkg/apis/meta/v1"
 	"testing"
```
reorder import

RealHarshThakur commented:
@akhilerm LGTM , I think there could be small things we could be missing as PR has gotten so huge but we can fix them later.

akhilerm commented:
> @akhilerm LGTM , I think there could be small things we could be missing as PR has gotten so huge but we can fix them later.

@RealHarshThakur can you make a list of those things, so that we can refer back to this PR, incase we need to update it later.

shubham14bajpai reviewed (approved):
LGTM

RealHarshThakur commented:
@akhilerm I meant I don't see anything wrong immediately. There could be no issues too :D

akhilerm reviewed (approved):
LGTM
